src/core_plugins/kibana/public/dashboard/panel/panel_header/panel_actions/build_context_menu.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * Loops through allActions and extracts those that belong on the given contextMenuPanelId
+ * @param {string} contextMenuPanelId
+ * @param {Array.<DashboardPanelAction>} allActions
+ */
+function getActionsForPanel(contextMenuPanelId, allActions) {
+  return allActions.filter(action => action.parentPanelId === contextMenuPanelId);
+}
+
+/**
+ * @param {String} contextMenuPanelId
+ * @param {Array.<DashboardPanelAction>} actions
+ * @param {Embeddable} embeddable
+ * @param {ContainerState} containerState
+ * @return {{
+ *   Array.<EuiContextMenuPanelItemShape> items - panel actions converted into the items expected to be on an
+ *     EuiContextMenuPanel,
+ *   Array.<EuiContextMenuPanelShape> childPanels - extracted child panels, if any actions also open a panel. They
+ *     need to be moved to the top level for EUI.
+ *  }}
+ */
+function buildEuiContextMenuPanelItemsAndChildPanels({ contextMenuPanelId, actions, embeddable, containerState }) {
+  const items = [];
+  const childPanels = [];
+  const actionsForPanel = getActionsForPanel(contextMenuPanelId, actions);
+  actionsForPanel.forEach(action => {
+    const isVisible = action.isVisible({ embeddable, containerState });
+    if (!isVisible) {
+      return;
+    }
+
+    let childPanelToOpenOnClick;
```

Comment:
I don't see a reason to have `childPanelToOpenOnClick`, since we're passing action into `convertPanelActionToContextMenuItem`. `convertPanelActionToContextMenuItem` can just check action for the right id.